pkg/agent/controller/noderoute/node_route_controller.go

```diff
@@ -406,8 +408,14 @@ func (c *Controller) deleteNodeRoute(nodeName string) error {
 }
 
 func (c *Controller) addNodeRoute(nodeName string, node *corev1.Node) error {
-	if _, installed, _ := c.installedNodes.GetByKey(nodeName); installed {
-		// Route is already added for this Node.
+	peerNodeMAC, err := getNodeMAC(node)
+	if err != nil {
+		return fmt.Errorf("error when retrieving MAC of Node %s: %v", nodeName, err)
+	}
+
+	nrInfo, installed, _ := c.installedNodes.GetByKey(nodeName)
+	if installed && nrInfo != nil && nrInfo.(*nodeRouteInfo).nodeMAC != nil && peerNodeMAC != nil && nrInfo.(*nodeRouteInfo).nodeMAC.String() == peerNodeMAC.String() {
```

Comment:
it feels like if they are both `nil`, we should also return early?

```golang
if installed && nrInfo != nil {
    installedMAC := nrInfo.(*nodeRouteInfo).nodeMAC
    if installedMAC == nil && peerNodeMAC == nil {
        return nil
    }
    if installedMAC != nil && peerNodeMAC != nil && installedMAC.String() == peerNodeMAC.String() { {
        return nil
    }
}
```

Reply:
It's fine to call `installedMAC.String()` when `installedMAC` is nil, it will get "".
```
type HardwareAddr []byte

func (a HardwareAddr) String() string {
	if len(a) == 0 {
		return ""
	}
	buf := make([]byte, 0, len(a)*3-1)
	for i, b := range a {
		if i > 0 {
			buf = append(buf, ':')
		}
		buf = append(buf, hexDigit[b>>4])
		buf = append(buf, hexDigit[b&0xF])
	}
	return string(buf)
}
```
So this can be simplified to:
```
if installed && nrInfo.(*nodeRouteInfo).nodeMAC.String() == peerNodeMAC.String()
```

Reply:
I am not sure. @tnqn please also take a look.

Reply:
We don't need to check `nrInfo != nil` and `peerNodeMAC != nil`?

Reply:
No, we don't need to check it. The "String" method checks the length of the struct. You can test it to confirm.

Reply:
Got it.